File: urbanairship-core/src/test/java/com/urbanairship/channel/ChannelApiClientTest.java
/* Copyright Airship and Contributors */

package com.urbanairship.channel;

import com.urbanairship.BaseTestCase;
import com.urbanairship.TestAirshipRuntimeConfig;
import com.urbanairship.TestRequestSession;
import com.urbanairship.config.AirshipUrlConfig;
import com.urbanairship.http.RequestBody;
import com.urbanairship.http.Response;

import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

public class ChannelApiClientTest extends BaseTestCase {

    private ChannelRegistrationPayload payload = new ChannelRegistrationPayload.Builder().build();
    private ChannelApiClient client;
    private TestRequestSession requestSession = new TestRequestSession();

    @Before
    public void setUp() {
        TestAirshipRuntimeConfig runtimeConfig = TestAirshipRuntimeConfig.newTestConfig();
        runtimeConfig.setUrlConfig(AirshipUrlConfig.newBuilder()
                                                   .setDeviceUrl("https://example.com")
                                                   .build());

        client = new ChannelApiClient(runtimeConfig, requestSession);
    }

    /**
     * Test create channel succeeds request if the status is 200
     */
    @Test
    public void testCreateChannelSucceedsRequest() throws Exception {
        requestSession.addResponse(200, "{ \"ok\": true, \"channel_id\": \"someChannelId\"}");

        Response<String> response = client.createChannelWithPayload(payload);

        assertEquals("POST", requestSession.getLastRequest().getMethod());
        assertEquals("https://example.com/api/channels/", requestSession.getLastRequest().getUrl().toString());
        assertEquals(requestSession.getLastRequest().getBody(), new RequestBody.Json(payload));
        assertEquals("someChannelId", response.getResult());
        assertEquals(200, response.getStatus());
    }

    /**
     * Test create channel request fails
     */
    @Test
    public void testCreateChannelFailsRequest() throws Exception {
        requestSession.addResponse(501);

        Response<String> response = client.createChannelWithPayload(payload);

        assertEquals("POST", requestSession.getLastRequest().getMethod());
        assertEquals("https://example.com/api/channels/", requestSession.getLastRequest().getUrl().toString());
        assertEquals(requestSession.getLastRequest().getBody(), new RequestBody.Json(payload));
        assertNull(response.getResult());
        assertEquals(501, response.getStatus());
    }

    /**
     * Test update channel succeeds request if the status is 200
     */
    @Test
    public void testUpdateChannelSucceedsRequest() throws Exception {
        requestSession.addResponse(200);

        Response<Void> response = client.updateChannelWithPayload("someChannelId", payload);

        assertEquals("PUT", requestSession.getLastRequest().getMethod());
        assertEquals("https://example.com/api/channels/someChannelId", requestSession.getLastRequest().getUrl().toString());
        assertEquals(requestSession.getLastRequest().getBody(), new RequestBody.Json(payload));
        assertEquals(200, response.getStatus());
    }

    /**
     * Test update channel request fails
     */
    @Test
    public void testUpdateChannelFailsRequest() throws Exception {
        requestSession.addResponse(501);

        String channelId = "someChannelId";
        Response<Void> response = client.updateChannelWithPayload(channelId, payload);

        assertEquals("PUT", requestSession.getLastRequest().getMethod());
        assertEquals("https://example.com/api/channels/someChannelId", requestSession.getLastRequest().getUrl().toString());
        assertEquals(requestSession.getLastRequest().getBody(), new RequestBody.Json(payload));
        assertEquals(501, response.getStatus());
    }

}

File: urbanairship-core/src/test/java/com/urbanairship/channel/ChannelApiClientTest.kt
/* Copyright Airship and Contributors */
package com.urbanairship.channel

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.urbanairship.TestAirshipRuntimeConfig
import com.urbanairship.TestRequestSession
import com.urbanairship.http.RequestAuth
import com.urbanairship.http.RequestBody
import com.urbanairship.remoteconfig.RemoteAirshipConfig
import com.urbanairship.remoteconfig.RemoteConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.StandardTestDispatcher
import kotlinx.coroutines.test.TestResult
import kotlinx.coroutines.test.resetMain
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.test.setMain
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@OptIn(ExperimentalCoroutinesApi::class)
@RunWith(AndroidJUnit4::class)
public class ChannelApiClientTest {

    private val payload = ChannelRegistrationPayload.Builder().build()
    private val config = TestAirshipRuntimeConfig()
    private val requestSession = TestRequestSession()
    private val testDispatcher = StandardTestDispatcher()

    private val client = ChannelApiClient(config, requestSession)

    @Before
    public fun setup() {
        Dispatchers.setMain(testDispatcher)
        config.updateRemoteConfig(
            RemoteConfig(
                airshipConfig = RemoteAirshipConfig(
                    deviceApiUrl = "https://example.com"
                )
            )
        )
    }

    @After
    public fun tearDown() {
        Dispatchers.resetMain()
    }

    @Test
    public fun testCreate(): TestResult = runTest {
        requestSession.addResponse(200, "{ \"ok\": true, \"channel_id\": \"someChannelId\"}")
        val response = client.createChannel(payload)
        val result = requireNotNull(response.value)

        assertEquals(
            Channel(
                "someChannelId", "https://example.com/api/channels/someChannelId"
            ), result
        )

        assertEquals("POST", requestSession.lastRequest.method)
        assertEquals(
            "https://example.com/api/channels/", requestSession.lastRequest.url.toString()
        )
        assertEquals(RequestAuth.GeneratedAppToken, requestSession.lastRequest.auth)
        assertEquals(requestSession.lastRequest.body, RequestBody.Json(payload))
    }

    @Test
    public fun testCreateNullUrl(): TestResult = runTest {
        config.updateRemoteConfig(RemoteConfig())
        requestSession.addResponse(200, "{ \"ok\": true, \"channel_id\": \"someChannelId\"}")
        val response = client.createChannel(payload)
        assertNotNull(response.exception)
    }

    @Test
    public fun testUpdate(): TestResult = runTest {
        requestSession.addResponse(200)
        val response = client.updateChannel("someChannelId", payload)
        val result = requireNotNull(response.value)

        assertEquals(
            Channel(
                "someChannelId", "https://example.com/api/channels/someChannelId"
            ), result
        )

        assertEquals("PUT", requestSession.lastRequest.method)
        assertEquals(RequestAuth.ChannelTokenAuth("someChannelId"), requestSession.lastRequest.auth)
        assertEquals(
            "https://example.com/api/channels/someChannelId", requestSession.lastRequest.url.toString()
        )
        assertEquals(requestSession.lastRequest.body, RequestBody.Json(payload))
    }

    @Test
    public fun testUpdateNullUrl(): TestResult = runTest {
        config.updateRemoteConfig(RemoteConfig())
        requestSession.addResponse(200, "{ \"ok\": true, \"channel_id\": \"someChannelId\"}")
        val response = client.updateChannel("someChannelId", payload)
        assertNotNull(response.exception)
    }
}
